apps/src/sites/studio/pages/maker/setup.js

```diff
@@ -31,38 +32,151 @@ const BoardSetupStatus = React.createClass({
     return `${baseFormURL}?${userAgentFieldFill}&${setupStatesFieldFill}`;
   },
 
+  detect() {
+    this.setState({...this.getInitialState(), isDetecting: true});
+
+    if (!isChrome()) {
+      this.fail('IsChrome');
+      return;
+    }
+
+    if (!isWindows()) {
+      this.hide('WindowsDrivers');
+    }
+
+    if (gtChrome33()) {
+      this.succeed('IsChrome');
+    } else {
+      this.fail('IsChrome');
+    }
+
+    const bc = new BoardController();
+    Promise.resolve()
+
+        // Is Chrome App Installed?
+        .then(() => this.spin('AppInstalled'))
+        .then(() => promiseWaitFor(200)) // Artificial delay feels better
+        .then(() =>
+            BoardController.ensureAppInstalled()
+                .then(() => this.succeed('AppInstalled'))
+                .catch(error => this.fail('AppInstalled')))
+
+        // Is board plugged in?
+        .then(() => this.spin('BoardPlug'))
+        .then(() => promiseWaitFor(200)) // Artificial delay feels better
+        .then(() =>
+            BoardController.getDevicePortName()
+                .then(() => this.succeed('BoardPlug'))
+                .catch(error => this.fail('BoardPlug')))
+
+        // Can we talk to the firmware?
+        .then(() => this.spin('BoardConnect'))
+        .then(() =>
+            bc.ensureBoardConnected()
+                .then(() => this.succeed('BoardConnect'))
+                .catch(error => this.fail('BoardConnect')))
+
+        // Can we initialize components successfully?
+        .then(() => this.spin('BoardComponents'))
+        .then(() => promiseWaitFor(200)) // Artificial delay feels better
+        .then(() => bc.connectWithComponents())
+        .then(() => this.celebrateAllSuccessful(bc))
+        .catch(error => this.fail('BoardComponents'))
+
+        // Put the board back in its original state, if possible
+        .then(() => bc.reset())
+        .then(() => this.setState({isDetecting: false}));
+  },
+
+  /**
+   * Play a song and animate some LEDs to demonstrate successful connection
+   * to the board.
+   * @param {BoardController} bc
+   * @returns {Promise} resolved when the song and animation are done.
+   */
+  celebrateAllSuccessful(bc) {
+    /**
+     * Run given function for each LED on the board in sequence, with givcen
+     * delay between them.
+     * @param {function(five.Led.RGB)} func
+     * @param {number} delay in milliseconds
+     * @returns {Promise} resolves after func is called for the last LED
+     */
+    function forEachLedInSequence(func, delay) {
+      return new Promise(resolve => {
+        const leds = bc.prewiredComponents.colorLeds;
+        let ledIndex = 0;
+        const interval = setInterval(() => {
+          func(leds[ledIndex]);
+          ledIndex++;
+          if (ledIndex >= leds.length) {
+            clearInterval(interval);
+            resolve();
+          }
+        }, delay);
```

Comment:
That might be even better than what I was thinking :-P

I had just been thinking of instead of doing a setInterval and clearing on the last index, do a setTimeout, and from within the function, call yourself again on a timeout unless it's the last index. Probably no better than what you have currently.